Add `machine.I2CTarget` implementing a generic I2C device

dpgeorge opened the pull request:
### Summary

(Edit: this summary was updated to reflect the most recent state of this PR.)

This PR implements a generic I2C target/peripheral/"slave" device.  It can work in two separate modes:
- a general device with interrupts/events/callbacks for low-level I2C operations like address match, read request and stop
- a memory device that allows reading/writing a specific region of memory (or "registers") on the target I2C device

The class is called `machine.I2CTarget`.  To make a memory device is very simple:
```py
from machine import I2CTarget

mem = bytearray(8)
i2c = I2CTarget(addr=67, mem=mem)
```
That's all that's needed to start the I2C target.  From then on it will respond to any I2C controller on the bus, allowing reads and writes to the `mem` bytearray.

You can also register to receive events.  For example to be notified when the memory is read/written:
```py
from machine import I2CTarget

def irq_handler(i2c_target):
    flags = i2c_target.irq().flags()
    if flags & I2CTarget.IRQ_END_READ:
        print("controller read from target starting at", i2c_target_memaddr)
    if flags & I2CTarget.IRQ_END_WRITE:
        print("controller wrote to target starting at", i2c_target_memaddr)

mem = bytearray(8)
i2c = I2CTarget(addr=67, mem=mem)
i2c.irq(irq_handler)
```

Instead of a memory device, an arbitrary I2C device can be implemented using all the events (see docs).

This is based on the discussion in #3935.

An implementation is provided for:
- rp2 (direct register access)
- alif (direct register access, same I2C hardware as rp2)
- stm32 (direct register access)
- esp32s2/esp32s3 (using the IDF, does not support all events)
- zephyr (using zephyr i2c_target API, all events supported)
- mimxrt
- samd

### Testing

An extmod test is added, and also a multitest.

The tests pass RPI_PICO, RPI_PICO2_W, PYBV10, PYBD_SF6, ALIF_ENSEMBLE.

### Trade-offs and Alternatives

This is a very simple implementation, but it works and is probably enough for most use cases.  There are lots of things that could be enhanced:
- [x] add Python callbacks to notify when the I2C controller reads/writes memory
- [x] implement 16-bit wide memory addressing (currently restricted to 8-bit addresses)
- [ ] implement other I2C targets, eg FIFO, or generic
- [ ] add support for `asyncio`, eg polling the device for events

dpgeorge committed:
stm32/i2c: Move I2C IRQ handlers from stm32_it.c to i2c.c.

And add MP_STATIC_ASSERT to statically check that the IRQ names are correct
on the MCU that it's compiled for.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
extmod/machine_i2c_target: Add new machine.I2CTarget class.

This commit implements a generic I2C target/peripheral/"slave" device,
called `machine.I2CTarget`.  It can work in two separate modes:

- A general device with interrupts/events/callbacks for low-level I2C
  operations like address match, read request and stop.

- A memory device that allows reading/writing a specific region of memory
  (or "registers") on the target I2C device.

To make a memory device is very simple:

    from machine import I2CTarget

    mem = bytearray(8)
    i2c = I2CTarget(addr=67, mem=mem)

That's all that's needed to start the I2C target.  From then on it will
respond to any I2C controller on the bus, allowing reads and writes to the
mem bytearray.

It's also possible to register to receive events.  For example to be
notified when the memory is read/written:

    from machine import I2CTarget

    def irq_handler(i2c_target):
        flags = i2c_target.irq().flags()
        if flags & I2CTarget.IRQ_END_READ:
            print("controller read target at addr", i2c_target.memaddr)
        if flags & I2CTarget.IRQ_END_WRITE:
            print("controller wrote target at addr", i2c_target.memaddr)

    mem = bytearray(8)
    i2c = I2CTarget(addr=67, mem=mem)
    i2c.irq(irq_handler)

Instead of a memory device, an arbitrary I2C device can be implemented
using all the events (see docs).

This is based on the discussion in #3935.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
stm32/machine_i2c_target: Implement I2CTarget class.

Works, tested on PYBV10, PYBD_SF2 and PYBD_SF6:

    buf = bytearray(16)
    machine.I2CTargetMemory("X", addr=67, mem=buf)

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
rp2/machine_i2c_target: Implement I2CTarget class.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
tests/extmod_hardware: Add self unittest for I2CTarget.

This test uses a SoftI2C controller wired to an I2CTarget on the one board,
and tests all functionality of the I2CTarget class.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
docs/library: Document the new machine.I2CTarget class.

With some working examples that show how to use all the features.

Signed-off-by: Damien George <damien@micropython.org>

chrismas9 commented:
Interesting timing. I have an immediate need for this. I have just written an implementation of https://github.com/micropython/micropython/issues/3935 in MicroPython. It implements callbacks and 16-bit addressing, but needs to be called regularly in the main loop to service I2C requests from the controller. It has some other limitations, but is enough for my applications.

I see this failed on STM32F091. I am using STM32L432. I will try it and report back.

Callbacks would be useful, but I can live without them. My inputs need to be debounced or filtered and that logic can update the read registers. For write registers I will maintain a backup copy and diff them periodically to update outputs.

Would the `memory_read `callback be called before the I2C reply so that the memory can be updated on demand, eg from a pin, ADC, etc?

Is it necessary to name it other than `I2CTarget `? If it were named `I2CTarget `now, with maybe `mode=memory` it would allow applications to keep working after a full featured `I2CTarget `was added to MicroPython.

kwagyeman commented:
@dpgeorge - Alif support? :)

robert-hh commented on ports/rp2/machine_i2c_target.c:
```diff
@@ -0,0 +1,185 @@
+#define IS_VALID_SDA(i2c, pin) (((pin) & 1) == 0 && (((pin) & 2) >> 1) == (i2c))
+
+static machine_i2c_target_data_t i2c_target_data[4];
+
```
It seems that the code only uses 2 of these data objects.

robert-hh commented:
I could start to implement this feature with the MIMXRT port. The NXP lib has support for the slave mode.

chrismas9 commented:
I would prefer callbacks to a flag as they allow for a much  faster response for I2C reads if the register is updated in the callback.  How would a flag work? Would the read function block until the flag is polled, the register updated and the flag is cleared, or would it immediately reply with stale data? The other problem, which I have with my python library, is that the Controller timeout has to be larger than the slowest poll time. I have use cases where such latency would be a problem, eg measuring motor current in a servo controller

chrismas9 commented:
I have added all STM32 families. I will submit the changes after testing them all on Eval boards.

chrismas9 commented:
I was hoping for a blocking read callback that could update a register with fresh data before the reply using clock stretching during the callback. I am making a series of modules which should behave like standard ICs. Most ADCs respond to a read with fresh data, either by clocking a SAR with SCL or using clock stretching until a conversion is finished. Any background loop that has to continuously update registers will result in reading data that is probably 10 to 100ms out of date. I'm not sure it's very useful to have a read callback after the event.

Write callbacks after a write are very useful as long as they contain the range of registers written.

robert-hh committed:
mimxrt/machine_i2c_target: Support I2C target mode.

The functionality is similar to the RP2 implementation.  The supported
address size is 7 bit.  In order to achieve a sufficient response, the
target I2C IRQ handler has to run from RAM, causing much more code moved to
RAM than required.

Tested with Teensy 4.1, MIMXRT1021EVK, MIMXRT1011EVK and MIMXRT1170, using
both a On-Board SoftI2C as controller and a RP2 Pico as external
controller.

Signed-off-by: Damien George <damien@micropython.org>
Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh commented:
Thanks. I have seen the test and will run it. And yes, I expect the API to change, but that can be adapted easily.

robert-hh commented:
There were some issues in the wrap tests, that are fixed now. After adapting `machine_i2c_target_memory.py` with the settings for this board, all tests pass. The code can still be optimized somewhat and tested at the various board types, especially the slower i.mx1011 variants.

dpgeorge commented:
> I was hoping for a blocking read callback that could update a register with fresh data before the reply using clock stretching during the callback.

This is an important question to discuss further.  With a read transaction (point of view from the controller) we have:
1. generate start condition
2. send address byte, with direction flag (read)
3. read byte from the target, so the target must drive SDA
4. generate stop condition

If the target is not ready in step 3 to send the data (eg an ADC that needs to be sampled) then the target could stretch the clock by holding SCL low.  Then release it when it has the new data ready.

If we want to support such clock stretching then the API and implementation will be quite different, because we need a way for the Python code to signal that it wants to stretch the clock (stall the bus), and then indicate when it is ready to unstall the bus.

There are real devices out there that do this clock stretching, eg ADS7128 (and it seems that device stretches SCL for quite some time when it's averaging many readings).  Other devices like ADS1115 require you to poll a register to see when the ADC conversion is ready.  And something like MCP3021 will do the ADC conversion during the first few bits of the data read (and those bits always read 0) so don't need to stretch SCL.

Looking at Zephyr's I2C target API, it does not support clock stretching, so it's not possible to implement a clock stretching target with Zephyr.

Do we need the MicroPython I2C target API to be able to do clock stretching?

Other considerations:
- clock stretching for writes to the target; this could be used for example to implement flow control when a lot of data is written out to the target (it holds SCL low if it's not ready to receive more data)
- support atomic reads from memory device, eg if the I2C controller is reading a 16-bit value from the I2C target, you don't want the I2C target to update that memory in the middle of the read or else the upper and lower bytes of the 16-bit value will be out of sync
- support atomic writes

Things can get complicated pretty quickly when looking at all these things.

kwagyeman commented:
Clock stretching is a very undesirable behavior of I2C.

Do not replicate that. It makes the bus unreliable as it means I2C transfers are accomplished in fixed time. This kills any MCUs ability to meet a schedule with an I2C shared with multiple device.

robert-hh commented:
>  Teensy already needs D0/D1 and D2/D3 and D11/D12 connected 
With the test of PAM and UART, these Pins are pairwise connected, D0 to D1, D2 to D3 and D11 to D12. For I2C, SDA (A4) and and SCL(A5)  have to be connected each to a GPIO pin. So I suggest to use D4 and D5. Could be as well A2 and A3. At the Teensy, SCL and SDA need an external pull-up resistor for a reliable test. The internal pull-up is not sufficiently strong.

robert-hh committed:
samd/machine_i2c_target: Support I2C target mode.

Supporting readfrom_mem*(). writeto_mem() and a set of IRQs.  Enabled by
default for SAMD51 devices and SAMD21 devices with external flash.

Tested with ItsyBitsy M4 and ItsyBitsy M0 with both on-board SoftI2C and a
RP2 Pico as controller.

Signed-off-by: Damien George <damien@micropython.org>
Signed-off-by: robert-hh <robert@hammelrath.com>

robert-hh commented:
I looked into other ports, whether they can support the I2CTargetMemory class. 
- The NRF port it looks promising. The library provides suitable functions. 
- ESP8266 does not provides I"C hardware support at all.
- The Renesas port seems to have I2C target mode hardware, but the HAL layer does not offer interfaces for using it.
- The ESP32 situation is crooked for several reasons.
 
  + The ESP32 API has three versions of I2C support. 1. The legacy API, which is currently used for I2C controller support. 2. The ng API Version 1. 3, The ng API version 2. Option b) is declared as getting expired with idf 6.x. I did not check yet for which IDF versions the ng API V2 is available. It is present in IDF v5.4. The legacy API and ng API must not be used at the same time. So if that is going to be used, the I2C class has to be updated.
  + Implementing the class  I2CTargetMemory for trial it turned out that it cannot properly support the target memory mode. The API offers two callback functions or receive and transmit. Each of them is called AFTER the respective bus transactions is finished. That makes it possible to support the writeto_mem() method properly, but fails to support readfrom_mem(). The only option which could be supported is getting the proper content from the previous state of the memory object. After it has been written, one has to call readfrom_mem() twice to get the proper content. It may be useful to have a I2CTargetMemory.write() method which fills the transmit buffer with data from a provided object.
  + One has to declare a suitable large temporary buffer. If on writeto_mem() the sent data is longer than the declared buffer, the amount of data actually provided to the callback is erratic. On readfrom_mem() the reading does not wrap. Instead for the excess data 0xff is sent.

dpgeorge commented:
Thanks @robert-hh for working on this for the other ports, that's very helpful!

Right now I'm reworking the API here to take into account the feedback, mainly to provide IRQs/callbacks for various I2C target events.  After that's done we can integrate your changes.

> I'm still unsure about the wrapping behavior for writing and reading. I could not find a source where this is defined as to be expected. Do you have a reference?

It's not really defined as a standard thing, it's just that a lot of devices will wrap around if you read beyond the end of their register range.  Eg I think the MMA on the PYBv1.x does this?

dpgeorge committed:
zephyr/machine_i2c_target: Implement I2CTarget class.

Tested and working on rpi_pico and nucleo_wb55rg.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
I've now updated this PR:
- renamed the class to `machine.I2CTarget`
- now supports creating either a memory device, or an arbitrary device
- now supports IRQs for various events

It's not finished but the included test is comprehensive and passes on stm32, rp2 and zephyr.

@robert-hh are you able to update the ports you worked on so far, to match the new code here?

robert-hh commented:
@dpgeorge Will do that.

robert-hh commented on docs/library/machine.I2CTarget.rst:
```diff
@@ -0,0 +1,78 @@
+
+    # Create an I2C target.  Depending on the port, extra parameters
+    # may be required to select the peripheral and/or pins to use.
+    i2c = I2CTargetMemory(addr=67, mem=mem)
```
Shouldn't that be: `i2c = I2CTarget(addr=67, mem=mem)`

dpgeorge replied:
Now fixed.

hoihu commented:
> Do we need the MicroPython I2C target API to be able to do clock stretching?

I would be in favour of that. 

If the "Target" device is an "intelligent device" it's very likely it has other tasks to do as well (usb, FS, network, sensor measurements...) -> It might not be ready all the time to serve the (read) interrupt since it's busy with other high priority tasks. We actually use such a setup in some of our products. I believe it's pretty common actually.

> support atomic reads from memory device, eg if the I2C controller is reading a 16-bit value from the I2C target, you don't want the I2C target to update that memory in the middle of the read or else the upper and lower bytes of the 16-bit value will be out of sync

oh yes, that's a concern I know of first hand ;) So yes - such a mutex or similar would be really nice to have. I tend to think of that as a shared RAM (via I2C, and directly on the MCU). With all it's pros and cons.

Other Comments
- blocking wait(s) on the controller side: would it be worth to add **async** API for read/write operations on the controller?
- Instead of "Target", perhaps use "Peripheral"? I2CPeripheralMemory?

dpgeorge commented:
> * The function `mp_machine_i2c_target_read_bytes()` can deal with values of len > 1, but not `mp_machine_i2c_target_write_bytes()`

I may need to change this to only deal with one byte at a time.  It depends on how it all turns out with the other ports.

> * is is as well acceptable to implement `mp_machine_i2c_target_read_bytes() ` with a length of 1?

Yes, that's fine to make this function deal only with one byte.

> * But when using i2c.writeto() and i2c.readfrom(), always one byte is skipped. for instance, with the object set tp "01234567", calling`i2c.readfrom(67, 1)` returns varying data with one value skipped

I will look into this.  But for now that should not impact the port-specific implementation.  As long as the port can pass the test that's added in this PR, it should be fully correct.

chrismas9 commented:
With regard to clock stretching I would still like to see an optional callback `update_mem=update_mem` called during clock stretching to update the buffer before it is read and sent to the controller. I would use this for fast updates such as reading a GPIO. This would also solve the atomic read problem as the driver would be stalled during the callback. If you don't want clock stretching use `update_mem=None`.

With regard to atomic multi-byte read and write what about adding the `readfrom_mem` and `writeto_mem` methods from `machine.I2C`? Buffer access method would be the same for controller and target and I2CTarget can handle the interleaving of reads and writes by controller and target.

robert-hh commented:
> With regard to clock stretching I would still like to see an optional callback update_mem=update_mem called during clock stretching to update the buffer before it is read and sent to the controller.

WIth the actual version there are quite a few IRQ events that can be set: `'IRQ_ADDR_MATCH', 'IRQ_END', 'IRQ_READ_REQ', 'IRQ_STOP', 'IRQ_WRITE_REQ'`. These could be used to update the buffer.

robert-hh commented:
@dpgeorge There seems to be a hiccup with the RP2 port and the value of data->status. When calling e.g. `i2c.readfrom_mem(67, 0, 8)` at the controller, and a callback is active for  IRQ_ADDR_MATCH, then this callback is  called for every single data byte, not only for the first. Timing below. The gap between the data bytes is the time, when the callback is active, and it shows b.t.w., that the RP2 uses clock stretching.

<img width="1597" height="375" alt="grafik" src="https://github.com/user-attachments/assets/699a91e7-b9c4-440f-8045-bb512c5f9b89" />

robert-hh commented on extmod/machine_i2c_target.c:
```diff
@@ -0,0 +1,356 @@
+    I2C_TARGET_IRQ_END = 1 << 3,
+    I2C_TARGET_IRQ_STOP = 1 << 4, // should be able to support this on all targets (rp2 needs custom code though), but maybe can't support addr_match?
+
+    I2C_TARGET_IRQ_MEM_ADDR_MATCH = 1 << 5, // needed?
```
These look not necessary. The ..START ones are at the same time that the ADDR_MATCH, and the ..END ones are related to STOP or END:

dpgeorge replied:
Unnecessary events have now been removed.

robert-hh commented on extmod/machine_i2c_target.c:
```diff
@@ -0,0 +1,356 @@
+            // TODO allow N bytes for address, with N=0 allowed
+            data->state = STATE_MEM_ADDR_SELECTED;
+            data->mem_addr = val % data->mem_len;
+            event(data, I2C_TARGET_IRQ_MEM_ADDR_MATCH, data->mem_addr, 0);
```
Seems to cause a duplicate call, because it follows the address byte handling, which already calls the event handler for I2C_TARGET_IRQ_MEM_ADDR_MATCH.

robert-hh commented on extmod/machine_i2c_target.c:
```diff
@@ -0,0 +1,356 @@
+        }
+    } else {
+        // Have a buffer.
+        if (data->state == STATE_MEM_ADDR_SELECTED) {
```
The additional calls for ADDR_MATCH events disappear, if the test is:
`        if (data->state == STATE_MEM_ADDR_SELECTED || data->state == STATE_IDLE) {`
This is the smallest change I could find, but may not be the best method to fix the state machine here.
But that seems to be only a problem of the RP2 port, because it does not have a specific ADDR_MATCH event.

robert-hh replied:
Just ` if (data->state ==  STATE_IDLE)` is sufficient.

dpgeorge committed:
alif/machine_i2c_target: Implement I2CTarget class.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
I have now implemented `I2CTarget` on the alif port.  The test included here passes, all IRQs are supported in hard IRQ mode.

dpgeorge commented:
> Interestingly, while upon send the Target uses clock-stretching, it does not do so on receive. But maybe there is a configuration which enables it upon FIFO overrun.

I tested clock stretching on the alif port.  The I2C hardware on alif has a 32-byte deep FIFO, for both TX and RX.  So you don't see clock stretching for controller -> target with just a few bytes being sent by the controller, because the alif target is queuing these bytes in its RX FIFO.  OTOH I did observe clock stretching for target -> controller: as long as the TX FIFO is empty the alif will stretch the clock.  So you don't need to respond immediately to `IRQ_READ_REQ`, you can delay that indefinitely and it'll stretch the clock until `i2c_target.write(...)` is called which fills the TX FIFO.

I guess all the ports are going to behave differently here, depending on whether they have a FIFO or not, and whether they can do clock stretching.

I'm not sure how useful these events/IRQs are going to be in practice to make real I2C targets/devices.  Still needs investigation.

robert-hh commented:
> I'm not sure how useful these events/IRQs are going to be in practice to make real I2C targets/devices. 

For me the ADDR_MATCH and STOP/END events seem to be the most useful ones. Maybe ADDR_MATCH could already be split into READ and WRITE. That is known already at tha time of address match. I2C_TARGET_IRQ_READ_REQ and I2C_TARGET_IRQ_WRITE_REQ seems less useful for being tested and triggered with every byte. That's a lot of overhead.

robert-hh commented on ports/rp2/machine_i2c_target.c:
```diff
@@ -106,8 +106,8 @@ static void i2c_target_handler(i2c_inst_t *i2c, i2c_slave_event_t event) {
-            //i2c_write_byte_raw(i2c, 0x12);
-            //break;
+            // i2c_write_byte_raw(i2c, 0x12);
+            // break;
```
Are these commented out lines needed? I guess not.

dpgeorge replied:
Removed.

robert-hh commented:
There is still the unexpected behavior that on reading from the Target the ADDR_MATCH handler is called for every byte.

robert-hh commented:
> I don't see this on the rp2 port. Using RPI_PICO_W,

Sorry. My old eyes. The second call is for I2C_TARGET_IRQ_MEM_ADDR_MATCH, not I2C_TARGET_IRQ_ADDR_MATCH. Anyhow I see the callback being called twice and data loss. But that's a different problem. It happens when sending data from a separate device at freq=400_000. With freq=100_000 all is fine. And of course it does not happen in the test script set-up, where the data is sent by SoftI2C and being at the same CPU, there cannot be any timing conflict between controller and target.

dpgeorge commented:
I have now reworked the rp2 I2CTarget implementation.  It's now the same as the alif one, and passes the included tests.

robert-hh commented:
Downloaded and tested with a RP2 Pico and a second board as Controller and a single callback for ADDR_MATCH. Runs fine at freq=100_000. Works fine with freq=400_000 for data requested from the slave. Behaves weird at freq=400_000 when data is sent from Controller:
- the callback is called twice instead of once.
- single bytes of the transfer are lost. Either the old data i still in the target memory at that place, or a NULL byte is inserted.
No improvement with that aspect compared to the previous version.
Without a callback data is transferred properly.

kwagyeman commented:
Hi Damien, will the current API be usable for replacing `pyb` in this case? https://github.com/openmv/openmv/blob/master/scripts/libraries/rpc.py#L502

It looks like fo get_bytes() I can use the read_into() in a loop while tracking time to receive data... However, for put_bytes(), I can't do the same kind of thing because mp_machine_i2c_target_write_bytes() seems to always write 1 byte even if there's no space in the FIFO on alif for example.

Can you adjust mp_machine_i2c_target_write_bytes() to return 0 if there's no space in the FIFO? Otherwise it looks like some type of interrupt driven thing is necessary to write data out.

dpgeorge commented:
> It looks like fo get_bytes() I can use the read_into() in a loop while tracking time to receive data...

Yes, you might be able to do that.  Although the FIFO is only 16 or 32 bytes deep, so if you expect a lot of data coming in, then it could overflow before you have a chance to read it out.

> Can you adjust mp_machine_i2c_target_write_bytes() to return 0 if there's no space in the FIFO? 

I tried this, but on rp2 and I assume also alif (they use the same Synopsis IP) the TX (outgoing) FIFO is flushed when there is an address match (also the RX FIFO).  So that means if you queue the data in the TX FIFO beforehand, it is all lost once the target is addressed by the controller.

To fix that either needs buffering outside the TX FIFO (ie in RAM), or use an interrupt based mechanism.

----

Maybe it's possible for you to use `I2CTarget` in memory device mode.  Just set an address size of 0 and a large memory buffer, eg:
```py
buf = bytearray(65536)
i2c_target = I2CTarget(0, addr, mem=buf, mem_addrsize=0)
```
Then for `get_bytes()` you just copy out of `buf`, and for `put_bytes()` you just copy into `buf`.  The actual transfer is then handled in the background.

Of course, that scheme would require having some way to know that new data has arrived in `buf`, and that data has been read out of `buf`.  Would having that capbility work for your case?

How much data do you have coming in and going out?  Do you need to be able to send large amounts of data from ROM (ie not copy to a temporary RAM buffer)?  What exactly is the protocol you use over I2C?

dpgeorge commented:
> Well, I need the method to transfer the requested data size or timeout. So, I'd need to know when the data has all been transferred within a certain time limit.

With the memory device, you'd set up the data buffer and then just let it do its thing.  If you need a timeout I guess you could do something like:
```py
def transfer_bytes(buf, timeout):
    i2c_target = I2CTarget(0, addr, mem=buf, mem_addrsize=0)
    t0 = time.ticks_ms()
    while True:
        if i2c_target.transfer_done():
            # controller read our buffer or wrote into our buffer
            i2c_target.deinit()
            return True
        if time.ticks_diff(time.ticks_ms(), t0) >= timeout:
            # timeout
            i2c_target.deinit()
            return False
        time.sleep(0)
```
That would handle both reads and writes.  The I2C target would disappear between transfers (I think that's how you have it at the moment?).

> The sizes of the buffers change all the time along with the timeouts.

What's the approximate maximum buffer size?  Is it OK for all buffers to be in RAM?

kwagyeman commented:
The largest buffer size can be 2^32 bytes. All would be in RAM or whatever is addressable via byte arrays or memory views.

For:

```
def transfer_bytes(buf, timeout):
    i2c_target = I2CTarget(0, addr, mem=buf, mem_addrsize=0)
    t0 = time.ticks_ms()
    while True:
        if i2c_target.transfer_done():
            # controller read our buffer or wrote into our buffer
            i2c_target.deinit()
            return True
        if time.ticks_diff(time.ticks_ms(), t0) >= timeout:
            # timeout
            i2c_target.deinit()
            return False
        time.sleep(0)
```

Could a method be added to machine class to do all this for you in C? Maybe add a poll method which you could select and wait on. It would return or timeout then. Each return could mark the completion of a I2C read/write.

dpgeorge commented:
> Could a method be added to machine class to do all this for you in C?

What's wrong with having it in Python, eg in a helper module in `micropython-lib`?  That would be more flexible than in C, and not any slower.

dpgeorge committed:
stm32/i2cslave: Change irq handler name to i2c_slave_irq_handler.

Remove the "ev" part, so this handler can be generalised to also handle
error IRQs.

Signed-off-by: Damien George <damien@micropython.org>

kwagyeman commented:
@dpgeorge - A helper in python is fine. Just thinking that the cleanest way to make it to have something that's pollable. https://docs.micropython.org/en/latest/library/select.html

Like, the way the OpenMV's RPC library works now is okay with a blocking loop... but, it needs to be rewritten eventually to support asyncio.

robert-hh commented:
>  Should be fixed now.

Confirmed. The above test passes at freq=400_000.

robert-hh commented:
Both the implementation for SAMD and MIMXRT are updated and rebased onto the actual state of this PR17365.

dpgeorge committed:
tests/multi_extmod: Add I2CTarget multi tests.

These require two boards wired together, SCL-SCL and SDA-SDA.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
esp32/machine_i2c_target: Implement I2CTarget class.

Only soft IRQs are supported.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
I've now added support for ESP32.  It was not easy, as @robert-hh discusses above in https://github.com/micropython/micropython/pull/17365#issuecomment-2979913776 :

> The ESP32 API has three versions of I2C support. 1. The legacy API, which is currently used for I2C controller support. 2. The ng API Version 1. 3, The ng API version 2. Option b) is declared as getting expired with idf 6.x. The ng API V2 is available starting version 5.4. So using is at the moment is not possible. The legacy API and ng API must not be used at the same time. So if that is going to be used, the I2C class has to be updated.

The esp32 port now uses IDF 5.4.2, so ng v2 I2C slave is available.  And there's an option `CONFIG_I2C_SKIP_LEGACY_CONFLICT_CHECK=y` which I used to allow it to co-exist with the existing I2C controller code.  That seems to work OK, at least as long as you only use I2C or I2CTarget (not both at the same time).

> Implementing the class I2CTargetMemory for trial it turned out that it cannot properly support the target memory mode. The API offers two callback functions or receive and transmit. Each of them is called AFTER the respective bus transactions is finished

Yes, that's pretty strange, and it confused me to start with.  I tried to get it working (by prefilling the TX FIFO before the transaction) but gave up when I realised that behaviour seems to only occur on ESP32 (original), not the other SoCs.  So I switched to development on ESP32S2 and it works much better, the transmit callback comes in before the transaction.  (It's still not ideal, but at least it can be made to work.)

> One has to declare a suitable large size for the temporary buffer

Yes, and for that I used the size of the user-supplied memory buffer.  This gets the tests working (most of them).

----

So ESP32 (ESP32S2 at least) now supports the memory buffer mode, and also the `IRQ_END_READ` and `IRQ_END_WRITE` events, as soft callbacks.  That means the code can be notified after the memory has been read/written.  And I added a `I2CTarget.memaddr()` function so the code can see which memory address was read/written.

I updated the documentation and other ports to work with the tweaked API / bindings.

The tests are also updated.

dpgeorge committed:
stm32/i2cslave: Account for slow addr_match callback.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
> About I2C V2. There was a PR #17696 this morning to update I2C support to V2.

That can be independent to the work here.  My testing on ESP32S2 shows that the existing I2C code works alongside this new I2C target code.

dpgeorge committed:
rp2/machine_i2c: Factor default pin macros to header file.

So they can be reused by the I2CTarget implementation.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
esp32/machine_i2c: Factor default pin macros to header file.

So the implementation of I2CTarget can use them.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
@projectgus I've now updated and tested this and it's ready to review.  Note that the commits still need to be squashed, but otherwise the code is in good shape.

@robert-hh there's a problem with mimxrt running the tests: it can do a read and a write transaction separately, but a read-then-write doesn't work, it doesn't respond to the write (ENODEV is raised).

robert-hh commented:
@dpgeorge I'll look into that. Which exact sequence caused the issue?

dpgeorge commented:
> Which exact sequence caused the issue?

When doing a `readfrom_mem(...)` on the controller side, when mimxrt is the target.  The controller will raise with ENODEV (works fine with `writeto_mem()` and with a simple `readfrom()`).

The included `tests/extmod_hardware/machine_i2c_target.py` test will show the issue.)

robert-hh commented:
> When doing a readfrom_mem(...) on the controller side

That is strange. I'm sure that `readfrom_mem(...)` and `writeto_mem` were my favorite test cases and that the test scripts `tests/extmod_hardware/machine_i2c_target.py` worked. I ran that test routinely.  I'll look into it.

projectgus reviewed (approved):
I think this will be really useful!

I haven't looked closely at all the port implementations but I think this is worth including, the main API looks sound (and flexible) and other details can be cleaned up later.

I left a couple of small comments, and agree with some of @robert-hh's outstanding ones as well - but they're all pretty minor.

robert-hh commented:
@dpgeorge I cannot replicate the problem. I fetched the last state from the repository and made a fresh build for a Teensy 4.1.  `readfrom_mem(...) ` works, and `tests/extmod_hardware/machine_i2c_target.py` runs fine as well:
```
micropython/tests$ mpremote run extmod_hardware/machine_i2c_target.py 
test_scan (__main__.TestMemory) ... ok
test_write (__main__.TestMemory) ... ok
test_write_wrap (__main__.TestMemory) ... ok
test_write_wrap_large (__main__.TestMemory) ... ok
test_read (__main__.TestMemory) ... ok
test_read_wrap (__main__.TestMemory) ... ok
test_read_wrap_large (__main__.TestMemory) ... ok
test_write_read (__main__.TestMemory) ... ok
test_write_read_read (__main__.TestMemory) ... ok
test_scan (__main__.TestMemoryIRQ) ... ok
test_write (__main__.TestMemoryIRQ) ... ok
test_read (__main__.TestMemoryIRQ) ... ok
test_scan (__main__.TestIRQ) ... ok
test_write (__main__.TestIRQ) ... ok
test_read (__main__.TestIRQ) ... ok
test_write_read (__main__.TestIRQ) ... ok
----------------------------------------------------------------------
Ran 16 tests

OK
```

```
micropython/tests$ ./run-tests.py -t a0 extmod_hardware/machine_i2c_target.py 
platform=mimxrt arch=armv7emdp inlineasm=thumb
pass  extmod_hardware/machine_i2c_target.py (unittest: 16 ran)
1 tests performed (16 individual testcases)
1 tests passed
```

dpgeorge commented on extmod/machine_i2c_target.c:
```diff
@@ -0,0 +1,107 @@
+typedef struct _machine_i2c_target_data_t {
+    uint8_t first_rx;
+    uint8_t addr;
+    uint8_t len; // stored as len-1 so it can represent up to 256
```
These are now 32-bit variables.

dpgeorge commented on ports/esp32/mpconfigport.h:
```diff
@@ -138,6 +138,9 @@
 #define MICROPY_PY_MACHINE_PWM_INCLUDEFILE  "ports/esp32/machine_pwm.c"
 #define MICROPY_PY_MACHINE_I2C              (1)
 #define MICROPY_PY_MACHINE_I2C_TRANSFER_WRITE1 (1)
+#define MICROPY_PY_MACHINE_I2C_TARGET       (SOC_I2C_SUPPORT_SLAVE && !CONFIG_IDF_TARGET_ESP32)
```
OK, done.

dpgeorge commented on ports/samd/machine_i2c.c:
```diff
@@ -72,16 +70,18 @@ typedef struct _machine_i2c_obj_t {
     uint8_t *buf;
 } machine_i2c_obj_t;
 
+uint8_t machine_i2c_pin_config(mp_obj_t pin_obj, uint8_t id, uint8_t pad_nr);
```
I've renamed this to `pin_config_for_i2c()` and put it in `pin_af.h`.

dpgeorge commented on ports/rp2/machine_i2c_target.c:
```diff
@@ -0,0 +1,185 @@
+
+static void mp_machine_i2c_target_memory_deinit(machine_i2c_target_memory_obj_t *self) {
+    i2c_slave_deinit(self->i2c_inst);
+}
```
Yes, you are right.  New fixed: the common extmod code clears out everything upon deinit and soft reset.

dpgeorge committed:
alif/machine_i2c: Allow changing I2C SCL/SDA pins.

Signed-off-by: Damien George <damien@micropython.org>

iabdalkader commented:
> @iabdalkader do you have pull-ups on SCL and SDA?

Yes, the BP has onboard pullups connected to 3.3v. I enabled them for all those tests.

> I tested it connected to a PYBD_SF2, with the PYBD_SF2 a controller and the ALIF_ENSEMBLE the I2C target.

Does a scan from PYBD_SF2 also return more than one address? Assuming the Ensemble board has nothing else on the bus, you should see just one address.

> It should be fine anyway, because it's exactly the same as the rp2 port, the silicon IP is the same (Design Ware / Synopsis). And rp2 also passes all tests.

While it is the same IP, we know that it's configured differently. For example, the STOP control is configured differently for AE3. Could that be related or some other IP config option that's affecting this?

I also tested another bus (i2c2) to be able to connect the probes directly instead of via jumpers, thought it might have something to do with, but it's the same. I don't know what else to try. Any suggestions?

dpgeorge commented:
> Yes, the BP has onboard pullups connected to 3.3v.

The AE3 is 1.8v, so you must be going through a level convertor.  Maybe that's the problem?

> Does a scan from PYBD_SF2 also return more than one address?

No, it returns just the single address.  I was using pins P3_7 and P3_6 on the Ensemble board.

> I don't know what else to try. Any suggestions?

Do you have an Ensemble board?  If so I suggest trying that because it's known to work (and don't use level shifters).  Try running `tests/extmod_hardware/machine_i2c_target.py`.  Be sure to connect P1_1 to P0_3, and P1_0 to P0_2.

You can also try a RPI_PICO connected to the Ensemble to run the multitest, `tests/multi_extmod/*.py`.

robert-hh commented:
@iabdalkader Did you use i2c.readfrom() or i2c.readfrom_mem()? You have to use the latter.

iabdalkader commented:
> @iabdalkader Did you use i2c.readfrom() or i2c.readfrom_mem()? You have to use the latter.

I use a bus pirate to read/write. I'm just not sure what the expected behavior here is, should it wrap around like that or restart the offset/pointer if it sees a stop?

iabdalkader commented:
@robert-hh That was the problem indeed. Thank you very much!

@dpgeorge Read/write is working fine, I was just using it wrong not writing the memory address. Scan however, still produces many addresses for me, not sure why.

kwagyeman commented:
Can confirm that Alif responds to the invalid I2C address 0x7C, and general call 0x00.

For general call, that's enabled by default to be responded to:

<img width="933" height="677" alt="image" src="https://github.com/user-attachments/assets/e38aa190-5de3-4b7c-99c2-e6a97ec95a5f" />

For 0x7C, this seems to be a reserved address but may be related to 10-bit addressing or some other feature of the IP that's enabled by default.

<img width="960" height="881" alt="image" src="https://github.com/user-attachments/assets/b79f23e8-7916-45e0-9d70-bea9a18bf4ed" />

dpgeorge commented:
> Tested with SAMD using using `tests/extmod_hardware/machine_i2c_target.py`. It locks up at TestPollling. Maybe the statement to skip it is missing in the test script, like it was added e.g. for mimxrt.

Thanks @robert-hh for testing on samd.  That TestPolling is intended to testing reading/writing the I2C FIFO outside the IRQ.  I'm happy for this test to fail, it's not a use case that we support yet, even if some ports can support it (eg because they have a relatively large I2C hardware FIFO, like rp2 and alif).

So I'll just skip this test on samd.  Can be fixed later if needed.

dpgeorge commented:
> Read/write is working fine, I was just using it wrong not writing the memory address

OK, thanks for confirming that.

> Can confirm that Alif responds to the invalid I2C address 0x7C, and general call 0x00.

I think that's acceptable.  These addresses are not usually used (in a 7-bit addressing system).  Responding to general call doesn't hurt.  If it turns out to be an issue, we can fix it then.

dpgeorge committed:
stm32/i2cslave: Add functions to read/write I2C data.

Instead of requiring the callback to consume/provide the data.  This allows
the data to be consumed/provided later on, which will stretch the I2C clock
until that occurs.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge committed:
stm32/i2cslave: Support i2c_slave_process_tx_end callback on F4.

The rounds out the F4 implementation to match the other supported MCUs.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
Commits have been squashed and docs improved.

kwagyeman commented:
Hi Damien, I tried to get our RPC library working. However, the AE3 seems to always crash. Is this code problematic?

```
class rpc_i2c_slave(rpc_slave):
    def __init__(self, slave_addr=0x12, i2c_bus=2):  # private
        self.__addr = slave_addr
        self.__bus = i2c_bus
        rpc_slave.__init__(self)
        self._stream_writer_queue_depth_max = 1

    def transfer(self, data, timeout_ms):
        view = memoryview(data)
        for i in range(0, len(view), 65535):
            target = machine.I2CTarget(self.__bus, self.__addr, mem=view[i : i + 65535], mem_addrsize=0)
            done = False
            def done_callback(trigger):
                nonlocal done
                done = True
            target.irq(done_callback)
            start = time.ticks_ms()
            while not done:
                time.sleep_ms(0)  # Is this needed??? Calls mp_event_handle_nowait() internally.
                if time.ticks_diff(time.ticks_ms(), start) >= timeout_ms:
                    target.deinit()
                    return None
            target.deinit()
        return view

    def get_bytes(self, buff, timeout_ms):  # protected
        buff = self.transfer(buff, timeout_ms)
        return buff

    def put_bytes(self, data, timeout_ms):  # protected
        self.transfer(data, timeout_ms)
```

Going to try the shadow register stuff next. But, the above should have worked for RPC. It seems to work about 6 times and then the system dies. Given it doesn't die on the first time it works... makes me think there's some memory corruption happening.

kwagyeman commented:
Found the issue, alif has infinite loops in their i2c_enable/disable:

<img width="490" height="412" alt="image" src="https://github.com/user-attachments/assets/49505bf0-00e4-4091-a5d4-f54ca54c25bc" />

Disable doesn't seem to go low in some cases after being told to.

kwagyeman commented:
They are both held low.

<img width="1138" height="549" alt="image" src="https://github.com/user-attachments/assets/1992dc3d-44ae-4fbb-b5cd-6dc978ea3fd8" />

And then if I zoom out:

<img width="1057" height="262" alt="image" src="https://github.com/user-attachments/assets/f4cd17ab-1a8b-4a6e-8e16-b10f27e100ee" />

The reference manual says to reset to the whole subsystem when this happens... 

<img width="1640" height="1964" alt="image" src="https://github.com/user-attachments/assets/9a7a0316-67d6-4bc9-bb9a-fe0d0fdfb8dc" />

The recovery bits mention they only work for master mode. When I looked into resetting I2C it appears it has a permanent clock and no reset lines... I guess I need to ask their app engineer for help.

kwagyeman commented:
Note that modifying/commenting out the disable while loop appears to break I2C across the whole system. Adding a timeout counter in the loop also appears to not work.
